File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/Principal.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk

import kalix.javasdk.{ Principal => JPrincipal }

/**
 * A principal associated with a request.
 */
sealed trait Principal

object Principal {

  /** Abstract principal representing all requests from the internet */
  case object Internet extends Principal

  /** Abstract principal representing all requests from self */
  case object Self extends Principal

  /** Abstract principal representing all requests from the backoffice */
  case object Backoffice extends Principal

  /**
   * A local service principal.
   */
  case class LocalService(name: String) extends Principal

  private[scalasdk] def toJava(principal: Principal): JPrincipal = principal match {
    case Internet           => JPrincipal.INTERNET
    case Self               => JPrincipal.SELF
    case Backoffice         => JPrincipal.BACKOFFICE
    case LocalService(name) => JPrincipal.localService(name)
  }

  private[scalasdk] def toScala(principal: JPrincipal): Principal = principal match {
    case JPrincipal.INTERNET                   => Internet
    case JPrincipal.SELF                       => Self
    case JPrincipal.BACKOFFICE                 => Backoffice
    case localService: JPrincipal.LocalService => LocalService(localService.getName)
  }

}

File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/Principal.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk;

import java.util.Objects;

/** A principal associated with a request. */
public interface Principal {

  /** Basic principals that have no additional configuration. */
  enum Basic implements Principal {
    INTERNET,
    SELF,
    BACKOFFICE
  }

  /** A local service principal. */
  final class LocalService implements Principal {
    private final String name;

    private LocalService(String name) {
      this.name = name;
    }

    public String getName() {
      return name;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      LocalService that = (LocalService) o;
      return Objects.equals(name, that.name);
    }

    @Override
    public int hashCode() {
      return Objects.hash(name);
    }

    public String toString() {
      return "LocalService(" + name + ")";
    }
  }

  /** Abstract principal representing all requests from the internet */
  Principal INTERNET = Basic.INTERNET;
  /** Abstract principal representing all requests from self */
  Principal SELF = Basic.SELF;
  /** Abstract principal representing all requests from the backoffice */
  Principal BACKOFFICE = Basic.BACKOFFICE;

  /** Create a local service principal with the given name */
  static LocalService localService(String name) {
    return new LocalService(name);
  }
}
